### PR TITLE
:sparkles: feat: TeamRegistry

### PR DESCRIPTION
**Pull request by Sabnock01**
## Description
Implements `TeamRegistry` (+ unit tests), a registry of teams for Curta Puzzles w/ the following actions:
- `batchRemoveMember(address calldata[] _members)`
- `batchSetApprovalForMember(address calldata[] _members, bool _approved)`
- `createTeam(address[] calldata _members)`
- `removeMember(address _member)`
- `setApprovalForMember(address _member, bool _approved)`
- `transferTeam(uint256 _from, uint256 _to)`
- `transferTeamLeadership(address _member)`

### CODE REVIEW

File: src/TeamRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+contract TeamRegistry {
```
Comment (fiveoutofnine) [inline src/TeamRegistry.sol:4]: let's add `title`/`notice`/`author` natspec

File: src/TeamRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    /// @notice Emitted when a team invite has been accepted.
+    /// @param  teamId The ID of the team.
+    /// @param  member The member of the team that accepted the invite.
+    event TeamInviteAccepted(uint32 teamId, address member);
+
+    /// @notice Emitted when team leadership is transferred.
+    /// @param  oldLeader The old leader of the team.
+    /// @param  newLeader The new leader of the team.
+    event TeamLeadershipTransferred(uint32 teamId, address oldLeader, address newLeader);
```
Comment (fiveoutofnine) [inline src/TeamRegistry.sol:50]: I've been using "present" tense for event names e.g. `AddPuzzle`, so let's keep that consistent here: `CreateTeam`, `AcceptTeamInvite`, `TransferTeamLeadership`

File: src/TeamRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    ///      `1`: Invited to team but not yet accepted
+    ///      `2`: Accepted invite and is team member
+    ///      `3`: Is team leader
+    mapping(uint32 => mapping(address => uint8)) public teamMemberStatus;
+
+    /// @notice Returns boolean if the address is a member of a team.
+    mapping(address => bool) public isTeamMember;
+
+    /// @notice Team ID counter. Starts at 1.
+    uint16 private teamId;
```
Comment (fiveoutofnine) [inline src/TeamRegistry.sol:68]: should this be `uint32`?

File: src/TeamRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    // -------------------------------------------------------------------------
+    // Functions
+    // -------------------------------------------------------------------------
+
+    /// @notice Create a team with a list of members.
+    /// @param  _members The team members to create the team with.
+    /// @dev    Reverts if `msg.sender` is already in a team.
+    function createTeam(address[] calldata _members) external {
+        // team leaders cannot create multiple teams
+        if (isTeamMember[msg.sender] == true) revert AlreadyInTeam(msg.sender);
```
Comment (fiveoutofnine) [inline src/TeamRegistry.sol:79]: agree w/ the comment: leaders shouldn't be able to create teams, but I think team members should be able to create teams (upon creating, they'd immediately leave whatever team they used to be in and become the leader). maybe we can just add `_currentTeamId` as a parameter to make this part easy

File: src/TeamRegistry.sol (hunk 3 above)
Comment (fiveoutofnine) [reply src/TeamRegistry.sol:68]: Screen Shot 2023-11-15 at 2 48 43 PM

Comment (Sabnock01) [reply src/TeamRegistry.sol:68]: Will make `uint256`

File: src/TeamRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        teamId += 1;
+        // make `msg.sender` team leader and say they are part of a team
+        teamMemberStatus[teamId][msg.sender] = 3;
+        isTeamMember[msg.sender] = true;
+        // send invites
+        for (uint8 i; i < _members.length;) {
+            // cannot create team with members of another team
+            if (isTeamMember[_members[i]] == true) {
+                revert AlreadyInTeam(_members[i]);
+            }
```
Comment (fiveoutofnine) [inline src/TeamRegistry.sol:90]: talked about this already, but no concern w/ sending out invites to addresses part of another team

**Comment by fiveoutofnine** [comment]
idk how to fix that fuzz test inconsistency

**Pull request merged**